<a href="https://colab.research.google.com/github/zzhining/ml_basic/blob/main/pandas_exercise_skeleton.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

np.random.seed(0)

# 과제 1
## Q1. 


<div class="alert alert-success">
[문제1]     
    
insurance.csv'파일 읽어서 df에 할당하고, 상위 5개 항목 출력하기
</div>

------------

[힌트]

pd.read_csv()

In [18]:

import pandas as pd
#data gapminder를 보고 DataFrame 만들기
#미리보기
df=pd.read_csv('./data/insurance.csv')


#print(df.info())
df.head(5)


age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520

## Q2. 


<div class="alert alert-success">
[문제2]     
    
df의 요약(평균, 4분위수, 전체 갯수 등)정보 확인하기
</div>

------------

In [4]:

import pandas as pd
#data gapminder를 보고 DataFrame 만들기
#미리보기
df=pd.read_csv('./data/insurance.csv')

df.info()



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1338 non-null   int64  
 1   sex       1338 non-null   object 
 2   bmi       1338 non-null   float64
 3   children  1338 non-null   int64  
 4   smoker    1338 non-null   object 
 5   region    1338 non-null   object 
 6   charges   1338 non-null   float64
dtypes: float64(2), int64(2), object(3)
memory usage: 73.3+ KB


## Q3. 


<div class="alert alert-success">
[문제3]     
    
신규 컬럼(double bmi)을 생성하고, bmi의 2배가 되는 값을 할당하기
</div>

------------

In [22]:
print(df)
#df.insert(loc=7, column='double_bmi', value=df['bmi']*2)
df

      age     sex     bmi  children smoker     region      charges  double_bmi
0      19  female  27.900         0    yes  southwest  16884.92400       55.80
1      18    male  33.770         1     no  southeast   1725.55230       67.54
2      28    male  33.000         3     no  southeast   4449.46200       66.00
3      33    male  22.705         0     no  northwest  21984.47061       45.41
4      32    male  28.880         0     no  northwest   3866.85520       57.76
...   ...     ...     ...       ...    ...        ...          ...         ...
1333   50    male  30.970         3     no  northwest  10600.54830       61.94
1334   18  female  31.920         0     no  northeast   2205.98080       63.84
1335   18  female  36.850         0     no  southeast   1629.83350       73.70
1336   21  female  25.800         0     no  southwest   2007.94500       51.60
1337   61  female  29.070         0    yes  northwest  29141.36030       58.14

[1338 rows x 8 columns]


age     sex     bmi  children smoker     region      charges  double_bmi
0      19  female  27.900         0    yes  southwest  16884.92400       55.80
1      18    male  33.770         1     no  southeast   1725.55230       67.54
2      28    male  33.000         3     no  southeast   4449.46200       66.00
3      33    male  22.705         0     no  northwest  21984.47061       45.41
4      32    male  28.880         0     no  northwest   3866.85520       57.76
...   ...     ...     ...       ...    ...        ...          ...         ...
1333   50    male  30.970         3     no  northwest  10600.54830       61.94
1334   18  female  31.920         0     no  northeast   2205.98080       63.84
1335   18  female  36.850         0     no  southeast   1629.83350       73.70
1336   21  female  25.800         0     no  southwest   2007.94500       51.60
1337   61  female  29.070         0    yes  northwest  29141.36030       58.14

[1338 rows x 8 columns]

## Q4. 


<div class="alert alert-success">
[문제4]     
    
신규 컬럼(debt)을 생성하고, 0으로 채우기
</div>

------------

In [23]:
df.insert(loc=8, column='debt', value=0)
df

age     sex     bmi  children smoker     region      charges  \
0      19  female  27.900         0    yes  southwest  16884.92400   
1      18    male  33.770         1     no  southeast   1725.55230   
2      28    male  33.000         3     no  southeast   4449.46200   
3      33    male  22.705         0     no  northwest  21984.47061   
4      32    male  28.880         0     no  northwest   3866.85520   
...   ...     ...     ...       ...    ...        ...          ...   
1333   50    male  30.970         3     no  northwest  10600.54830   
1334   18  female  31.920         0     no  northeast   2205.98080   
1335   18  female  36.850         0     no  southeast   1629.83350   
1336   21  female  25.800         0     no  southwest   2007.94500   
1337   61  female  29.070         0    yes  northwest  29141.36030   

      double_bmi  debt  
0          55.80     0  
1          67.54     0  
2          66.00     0  
3          45.41     0  
4          57.76     0  
...          ...   ...  
1333       61.94     0  
1334       63.84     0  
1335       73.70     0  
1336       51.60     0  
1337       58.14     0  

[1338 rows x 9 columns]

## Q5. 


<div class="alert alert-success">
[문제5]     
    
debt에 0부터 값이 증가하도록 채우기
</div>

------------

In [25]:
#df['debt'] = range(1, len(df) + 1)
df

age     sex     bmi  children smoker     region      charges  \
0      19  female  27.900         0    yes  southwest  16884.92400   
1      18    male  33.770         1     no  southeast   1725.55230   
2      28    male  33.000         3     no  southeast   4449.46200   
3      33    male  22.705         0     no  northwest  21984.47061   
4      32    male  28.880         0     no  northwest   3866.85520   
...   ...     ...     ...       ...    ...        ...          ...   
1333   50    male  30.970         3     no  northwest  10600.54830   
1334   18  female  31.920         0     no  northeast   2205.98080   
1335   18  female  36.850         0     no  southeast   1629.83350   
1336   21  female  25.800         0     no  southwest   2007.94500   
1337   61  female  29.070         0    yes  northwest  29141.36030   

      double_bmi  debt  
0          55.80     1  
1          67.54     2  
2          66.00     3  
3          45.41     4  
4          57.76     5  
...          ...   ...  
1333       61.94  1334  
1334       63.84  1335  
1335       73.70  1336  
1336       51.60  1337  
1337       58.14  1338  

[1338 rows x 9 columns]

## Q6. 


<div class="alert alert-success">
[문제6]     
    
1번 인덱스의 정보 출력
</div>

------------

[힌트]
iloc



age                  18  
sex                male  
bmi               33.77  
children              1  
smoker               no  
region        southeast  
charges         1725.55  
double bmi        67.54  
debt                  1  

In [26]:
df.iloc[1]

age                  18
sex                male
bmi               33.77
children              1
smoker               no
region        southeast
charges       1725.5523
double_bmi        67.54
debt                  2
Name: 1, dtype: object

## Q7. 


<div class="alert alert-success">
[문제7]     
    
'double bmi'열 삭제하기
</div>

In [ ]:
#df.columns.get_loc('double_bmi') #7열 반환
#df.drop('double_bmi', axis=1, inplace=True)  axis=1 이여야 열 삭제 axis=0이면 행삭제
df

age     sex     bmi  children smoker     region      charges  \
0      19  female  27.900         0    yes  southwest  16884.92400   
1      18    male  33.770         1     no  southeast   1725.55230   
2      28    male  33.000         3     no  southeast   4449.46200   
3      33    male  22.705         0     no  northwest  21984.47061   
4      32    male  28.880         0     no  northwest   3866.85520   
...   ...     ...     ...       ...    ...        ...          ...   
1333   50    male  30.970         3     no  northwest  10600.54830   
1334   18  female  31.920         0     no  northeast   2205.98080   
1335   18  female  36.850         0     no  southeast   1629.83350   
1336   21  female  25.800         0     no  southwest   2007.94500   
1337   61  female  29.070         0    yes  northwest  29141.36030   

      double_bmi  debt  
0          55.80     1  
1          67.54     2  
2          66.00     3  
3          45.41     4  
4          57.76     5  
...          ...   ...  
1333       61.94  1334  
1334       63.84  1335  
1335       73.70  1336  
1336       51.60  1337  
1337       58.14  1338  

[1337 rows x 9 columns]

## Q8. 


<div class="alert alert-success">
[문제8]     
    
 'age'열, 'sex'열만 출력하기
</div>

In [37]:
df[['age','sex']]

age     sex
0      19  female
1      18    male
2      28    male
3      33    male
4      32    male
...   ...     ...
1333   50    male
1334   18  female
1335   18  female
1336   21  female
1337   61  female

[1337 rows x 2 columns]

# 과제2
## Q9.
<div class="alert alert-success">
[문제9]     
    
나이를 1살씩 증가시키기
</div>

In [41]:
#df['age']=df['age']+1
df

age     sex     bmi  children smoker     region      charges  \
0      20  female  27.900         0    yes  southwest  16884.92400   
1      19    male  33.770         1     no  southeast   1725.55230   
2      29    male  33.000         3     no  southeast   4449.46200   
3      34    male  22.705         0     no  northwest  21984.47061   
4      33    male  28.880         0     no  northwest   3866.85520   
...   ...     ...     ...       ...    ...        ...          ...   
1333   51    male  30.970         3     no  northwest  10600.54830   
1334   19  female  31.920         0     no  northeast   2205.98080   
1335   19  female  36.850         0     no  southeast   1629.83350   
1336   22  female  25.800         0     no  southwest   2007.94500   
1337   62  female  29.070         0    yes  northwest  29141.36030   

      double_bmi  debt  
0          55.80     1  
1          67.54     2  
2          66.00     3  
3          45.41     4  
4          57.76     5  
...          ...   ...  
1333       61.94  1334  
1334       63.84  1335  
1335       73.70  1336  
1336       51.60  1337  
1337       58.14  1338  

[1337 rows x 9 columns]

## Q10.
<div class="alert alert-success">
[문제10]     
    
charges의 표시형식을 소수점 아래 2자리로 바꾸기
</div>

[힌트]

함수 생성, 적용(apply/map)

In [42]:
df

age     sex     bmi  children smoker     region      charges  \
0      20  female  27.900         0    yes  southwest  16884.92400   
1      19    male  33.770         1     no  southeast   1725.55230   
2      29    male  33.000         3     no  southeast   4449.46200   
3      34    male  22.705         0     no  northwest  21984.47061   
4      33    male  28.880         0     no  northwest   3866.85520   
...   ...     ...     ...       ...    ...        ...          ...   
1333   51    male  30.970         3     no  northwest  10600.54830   
1334   19  female  31.920         0     no  northeast   2205.98080   
1335   19  female  36.850         0     no  southeast   1629.83350   
1336   22  female  25.800         0     no  southwest   2007.94500   
1337   62  female  29.070         0    yes  northwest  29141.36030   

      double_bmi  debt  
0          55.80     1  
1          67.54     2  
2          66.00     3  
3          45.41     4  
4          57.76     5  
...          ...   ...  
1333       61.94  1334  
1334       63.84  1335  
1335       73.70  1336  
1336       51.60  1337  
1337       58.14  1338  

[1337 rows x 9 columns]

In [44]:
df = df.astype({'charges': float}).round(2)
df

age     sex    bmi  children smoker     region   charges  double_bmi  \
0      20  female  27.90         0    yes  southwest  16884.92       55.80   
1      19    male  33.77         1     no  southeast   1725.55       67.54   
2      29    male  33.00         3     no  southeast   4449.46       66.00   
3      34    male  22.70         0     no  northwest  21984.47       45.41   
4      33    male  28.88         0     no  northwest   3866.86       57.76   
...   ...     ...    ...       ...    ...        ...       ...         ...   
1333   51    male  30.97         3     no  northwest  10600.55       61.94   
1334   19  female  31.92         0     no  northeast   2205.98       63.84   
1335   19  female  36.85         0     no  southeast   1629.83       73.70   
1336   22  female  25.80         0     no  southwest   2007.94       51.60   
1337   62  female  29.07         0    yes  northwest  29141.36       58.14   

      debt  
0        1  
1        2  
2        3  
3        4  
4        5  
...    ...  
1333  1334  
1334  1335  
1335  1336  
1336  1337  
1337  1338  

[1337 rows x 9 columns]

## Q11.
<div class="alert alert-success">
[문제11]     
    
컬럼 정렬(알파벳 순서대로) : age, bmi, charges... 순으로 정렬
</div>

[Hint]
sort_index()


In [5]:
df_sorted = df.sort_index(axis=1)
print(df_sorted)

      age     bmi      charges  children     region     sex smoker
0      19  27.900  16884.92400         0  southwest  female    yes
1      18  33.770   1725.55230         1  southeast    male     no
2      28  33.000   4449.46200         3  southeast    male     no
3      33  22.705  21984.47061         0  northwest    male     no
4      32  28.880   3866.85520         0  northwest    male     no
...   ...     ...          ...       ...        ...     ...    ...
1333   50  30.970  10600.54830         3  northwest    male     no
1334   18  31.920   2205.98080         0  northeast  female     no
1335   18  36.850   1629.83350         0  southeast  female     no
1336   21  25.800   2007.94500         0  southwest  female     no
1337   61  29.070  29141.36030         0  northwest  female    yes

[1338 rows x 7 columns]


## Q12.
<div class="alert alert-success">
[문제12]     
    
값 정렬(age) : 나이 순으로 정렬
</div>

[Hint]
sort_values


In [6]:
df2 = df.sort_values(by='age')
df2

age     sex     bmi  children smoker     region       charges
57     18    male  31.680         2    yes  southeast  34303.167200
31     18  female  26.315         0     no  northeast   2198.189850
1023   18    male  23.320         1     no  southeast   1711.026800
427    18  female  29.165         0     no  northeast   7323.734819
1317   18    male  53.130         0     no  southeast   1163.462700
...   ...     ...     ...       ...    ...        ...           ...
603    64  female  39.050         3     no  southeast  16085.127500
534    64    male  40.480         0     no  southeast  13831.115200
1241   64    male  36.960         2    yes  southeast  49577.662400
801    64  female  35.970         0     no  southeast  14313.846300
768    64  female  39.700         0     no  southwest  14319.031000

[1338 rows x 7 columns]

## Q13.
<div class="alert alert-success">
[문제13]     
    
나이('age') 평균구하기
</div>

[Hint]
mean

In [7]:
age=df['age'].mean()
print(age)

39.20702541106129


## Q14.
<div class="alert alert-success">
[문제14]     
    
보험료('charges')와 가장 상관관계가 높은 변수는?
</div>

[Hint]
 - 공분산: df.cov()
 - 상관계수: df.corr()

In [47]:
columns = list(df.columns)
 
num_df = df.select_dtypes(include='number')

# 상관계수 행렬
corr_matrix = num_df.corr()

# 공분산 행렬
cov_matrix = num_df.cov()

print("상관계수 행렬:")
print(corr_matrix)

print("\n공분산 행렬:")
print(cov_matrix)


상관계수 행렬:
               age       bmi  children   charges
age       1.000000  0.109272  0.042469  0.299008
bmi       0.109272  1.000000  0.012759  0.198341
children  0.042469  0.012759  1.000000  0.067998
charges   0.299008  0.198341  0.067998  1.000000

공분산 행렬:
                   age           bmi    children       charges
age         197.401387      9.362337    0.719303  5.087480e+04
bmi           9.362337     37.187884    0.093795  1.464730e+04
children      0.719303      0.093795    1.453213  9.926742e+02
charges   50874.802298  14647.304426  992.674197  1.466524e+08


# 과제 3
## Q15.
<div class="alert alert-success">
[문제15]     
    
 남자, 여자가 각각 몇명있는지 구하기
</div>

[Hint]
pd.value_counts


In [9]:

female_count = df["sex"].value_counts()["female"]
male_count = df["sex"].value_counts()["male"]
print(f'남:{female_count}, 여:{male_count}')


남:662, 여:676


## Q16.
<div class="alert alert-success">
[문제16]     
    
자녀(children)가 3명인 데이터 출력하기
</div>

In [10]:
df_filtered = df[df["children"] == 3]
print(df_filtered)


      age     sex     bmi  children smoker     region      charges
2      28    male  33.000         3     no  southeast   4449.46200
7      37  female  27.740         3     no  northwest   7281.50560
25     59  female  27.720         3     no  southeast  14001.13380
36     62  female  32.965         3     no  northwest  15612.19335
54     40  female  28.690         3     no  northwest   8059.67910
...   ...     ...     ...       ...    ...        ...          ...
1301   62    male  30.875         3    yes  northwest  46718.16325
1314   30  female  23.655         3    yes  northwest  18765.87545
1320   31    male  31.065         3     no  northwest   5425.02335
1332   52  female  44.700         3     no  southwest  11411.68500
1333   50    male  30.970         3     no  northwest  10600.54830

[157 rows x 7 columns]


## Q16.
<div class="alert alert-success">
[문제16]     
    
자녀(children)가  5명 이상인 데이터 출력하기
</div>

In [11]:

df_filtered = df[df["children"] >= 5]
print(df_filtered)

      age     sex     bmi  children smoker     region      charges
32     19  female  28.600         5     no  southwest   4687.79700
71     31    male  28.500         5     no  northeast   6799.45800
166    20  female  37.000         5     no  southwest   4830.63000
413    25    male  23.900         5     no  southwest   5080.09600
425    45    male  24.310         5     no  southeast   9788.86590
438    52  female  46.750         5     no  southeast  12592.53450
568    49  female  31.900         5     no  southwest  11552.90400
640    33    male  42.400         5     no  southwest   6666.24300
877    33    male  33.440         5     no  southeast   6653.78860
932    46    male  25.800         5     no  southwest  10096.97000
937    39  female  24.225         5     no  northwest   8965.79575
969    39  female  34.320         5     no  southeast   8596.82780
984    20    male  30.115         5     no  northeast   4915.05985
1085   39  female  18.300         5    yes  southwest  19023.2

## Q17.
<div class="alert alert-success">
[문제17]     
    
 bmi가  17 미만인 데이터 출력하기
</div>

In [12]:

df_filtered = df[df["bmi"] <= 17]
print(df_filtered)

      age     sex     bmi  children smoker     region     charges
172    18    male  15.960         0     no  northeast  1694.79640
428    21  female  16.815         1     no  northeast  3167.45585
1226   38    male  16.815         2     no  northeast  6640.54485


# 과제 4 
## Q18.
<div class="alert alert-success">
[문제18]     
    
성별에 따른 보험료 평균
</div>

In [13]:
avg_charges = df.groupby("sex")["charges"].mean()
print(avg_charges)


sex
female    12569.578844
male      13956.751178
Name: charges, dtype: float64


## Q19.
<div class="alert alert-success">
[문제19]     
    
흡연여부에 따른 보험료 평균
</div>

In [14]:
avg_charges = df.groupby("smoker")["charges"].mean()
print(avg_charges)


smoker
no      8434.268298
yes    32050.231832
Name: charges, dtype: float64


## Q20.
<div class="alert alert-success">
[문제20]     
    
지역별 보험료 평균
</div>

In [15]:

avg_charges = df.groupby("region")["charges"].mean()
print(avg_charges)

region
northeast    13406.384516
northwest    12417.575374
southeast    14735.411438
southwest    12346.937377
Name: charges, dtype: float64


## Q21.
<div class="alert alert-success">
[문제21]     
    
흡연여부 + 지역별 데이터 평균
</div>

In [16]:
avg_charges = df.groupby(["region","smoker"])["charges"].mean()
print(avg_charges)

region     smoker
northeast  no         9165.531672
           yes       29673.536473
northwest  no         8556.463715
           yes       30192.003182
southeast  no         8032.216309
           yes       34844.996824
southwest  no         8019.284513
           yes       32269.063494
Name: charges, dtype: float64


## Q22.
<div class="alert alert-success">
[문제22]     
    
 흡연여부 + 지역별 데이터 평균에서 'charges' 만 츨력하기
</div>

In [66]:
print(avg_charges.values)
list2=avg_charges.values
list2 = list2.reshape(-1,1)
print(list2)


[ 9165.53167171 29673.53647284  8556.46371521 30192.00318241
  8032.21630894 34844.99682363  8019.28451307 32269.06349362]
[[ 9165.53167171]
 [29673.53647284]
 [ 8556.46371521]
 [30192.00318241]
 [ 8032.21630894]
 [34844.99682363]
 [ 8019.28451307]
 [32269.06349362]]


## Q23.
<div class="alert alert-success">
[문제23]     
    
 흡연여부 + 지역별 데이터 크기(size)
</div>

## Q24.
<div class="alert alert-success">
[문제24]     
    
smoker 여부에 따라, group_val 컬럼에 '흡연자', '비흡연자' 표시하기
</div>